source/manual/datalabs-bigquery.html.md

```diff
@@ -0,0 +1,86 @@
+
+## Google BigQuery best practice
+
+When using Google BigQuery you should:
+
+- minimise costs
+- optimise performance
+
+### Minimise Google BigQuery costs
+
+Google BigQuery charges based on the amount of data queried. You should try to minimise the amount of data you query to minimise costs.
+
+You should not use `SELECT *` in your queries if possible, as you could potentially be querying many unnecessary columns of data. Also, you should make sure you do not [query too many tables when using wildcards](https://cloud.google.com/bigquery/docs/querying-wildcard-tables).
+
+Use the query validator to the bottom right of the editor in the console to check query costs before running that query. For API queries, for example using Python, consider [using a dry run step to validate the query costs before running the code](https://cloud.google.com/bigquery/docs/best-practices-costs#performing_a_dry_run).
+
+If you are developing queries, consider using an intraday table. It has the same schema as the main `ga_sessions_YYYYMMDD` tables, but has less data stored for most of the day which makes queries cheaper to run.
+
+Be aware that intraday tables may not be representative behaviour. Queries using intraday tables are also likely to break the next day, as UA automatically moves the intraday table.
+
+See the [introduction to controlling BigQuery costs](https://cloud.google.com/bigquery/docs/controlling-costs) for more information.
+
+### Optimise Google BigQuery performance
+
+For complex queries, Google BigQuery automatically scales resources without extra charges to us. However, complex queries can take a long time to run, and even fail if Google BigQuery cannot scale enough resources.
+
+To mitigate these issues and to speed up analysis, you should optimise your queries before running them.
+
+See the [introduction to optimising query performance](https://cloud.google.com/bigquery/docs/best-practices-performance-overview) for more information.
+
+### Other best practice
+
+You should consider deleting datasets that you no longer use as we are charged a small fee for data storage. This also helps make sure you do not run queries incorrectly against old data.
+
+Consider creating a dataset for your own personal use. For example, use your personal dataset when conducting peer reviews so you do not overwrite live data. You must set the location to `EU`. You could set a table expiry condition to delete tables after a certain number of days.
+
+See the [BigQuery documentation on creating datasets](https://cloud.google.com/bigquery/docs/datasets) for more information.
+
+### Team-developed tools for Google BigQuery
+
+The GOV.UK Data Labs team has developed some tools to access Google BigQuery. These tools are for users outside of the team with limited SQL experience. One such tool is [alphagov/modular_sql](https://github.com/alphagov/modular_sql), a lightweight pipeline to combine multiple SQL scripts and generate Google BigQuery tables.
```

Comment:
There's an older data processing pipeline / repo that we developed that is useful for A/B testing analyis.

The repo itself is old and no longer maintained but has some useful queries to learn from. However, the modular_sql repo might be a better starting point.

I'd still recommend referencing https://github.com/alphagov/govuk-network-data as it can be useful for A/B testing and creating structured data at the session level.